source/simulator/core.cc

```diff
@@ -2241,6 +2256,297 @@ namespace aspect
           break;
         }
 
+        case NonlinearSolver::Newton_Stokes:
+        {
+          if (parameters.free_surface_enabled)
+            free_surface->execute ();
+
+          double initial_temperature_residual = 0;
+          double initial_stokes_residual      = 0;
+          std::vector<double> initial_composition_residual (parameters.n_compositional_fields,0);
+
+          //parameters.nonlinear_iteration_number = 0;
+
+          double max = std::numeric_limits<double>::max();
+
+          double initial_newton_residual_velo = 0;
+          double initial_newton_residual_pres = 0;
+          double initial_newton_residual = 0;
+
+          double newton_residual_velo = 0;
+          double newton_residual_pres = 0;
+          double newton_residual = 0;
+          double newton_residual_old = 0;
+
+          parameters.switch_initial_newton_residual = 1;
+          parameters.newton_residual = 1;
+          parameters.newton_theta = 0;
+
+          const double piccard_linear_stokes_solver_tolerance = parameters.linear_stokes_solver_tolerance;
+
+          bool   use_picard = true;
+
+          for (unsigned int i=0; (! ((i >= parameters.max_nonlinear_iterations) // regular timestep
+                                     ||
+                                     ((pre_refinement_step < parameters.initial_adaptive_refinement) // pre-refinement
+                                      &&
+                                      (i >= parameters.max_nonlinear_iterations_in_prerefinement)))); ++i)
+            {
+              assemble_newton_stokes_system = true;
+
+
+              double stokes_residual = 0;
+
+              solution.block(introspection.block_indices.pressure) = 0;
+              solution.block(introspection.block_indices.velocities) = 0;
+
+              pcout << "Newton iteration number: " << nonlinear_iteration << std::endl;
```

Comment:
Usually it is fine to do a iterated stokes type of solve (solve temp, solve comp, iterate Stokes) even when your material is nonlinear dependent on temperature and composition. But I have noticed sometimes that the iterated IMPES responds differently. I have based this Newton solver on iterated IMPES because it I think it is technically the best, but iterated Stokes might be faster.